### PR TITLE
[CORE-1352] WIP Document upcoming HTTPS redirect feature

### PR DESCRIPTION
**Pull request by greg-1-anderson**
Closes #n/a

## Effect
PR includes the following changes:
- Document `enforce_ttps` configuration setting on [Pantheon YAML Configuration Files](https://pantheon.io/docs/pantheon-yml/)
- Describes how to set up HTTPS redirects on [Switching Sites from HTTP to HTTPS](https://pantheon.io/docs/http-to-https/)
- Remove outdated docs/hsts 

[edited by @ari-gold]

## Remaining Work
- [x] Update HSTS documentation page
- [x] Update Pantheon YML documentation page
- [x] Document behavior of hsts header in https mode (min duration) and hsts mode (long duration with subdomains and preload)
- [x] Engineering review
- [x] Deploy HTTPS/HSTS feature to production
- [ ] Product owner sign-off (@ari-gold)
- [ ] Documentation review

## Post Launch
To be completed by the docs team upon merge: 
- [ ] Redirect `/docs/hsts to `/docs/pantheon-yml#enforce-https-+-hsts`
- [ ] Include/exclude pages ^ respectively within docs search service provider (if applicable)
- [ ] Update Status Report
- [ ] Archive from **Done** in Waffle

### CODE REVIEW

File: source/_docs/hsts.md [hunk 1]
```diff
@@ -3,15 +3,15 @@ title: Require HTTPS with the HSTS Header
...
 
 <div class="alert alert-info" role="alert">
   <h4 class="info">Note</h4>
   <p markdown="1">Before adding HSTS to your site, you should review and understand the configuration options that are available. A HSTS header that is configured incompletely or not securely enough diminishes the security protection that HSTS provides.</p>
 </div>
 
+Adding HSTS to your site may be done automatically via the [`scheme` setting in your pantheon.yml file](https://pantheon.io/docs/pantheon-yml/). If you need more control than offered by the built-in feature, then you may manually add HSTS to your site by using a module or plugin, as described below.
+
 ## Deploy and Configure a HSTS Header by Module or Plugin
```
Comment (ari-gold) [inline source/_docs/hsts.md]: Let's remove the section on module and plugin, or add a qualifier... "If for some reason you'd like finer grain control of HSTS than offered via pantheon.yml, then you can..."

File: source/_docs/hsts.md [hunk 2]
```diff
@@ -3,15 +3,15 @@ title: Require HTTPS with the HSTS Header
...
 tags: [security]
 ---
-After you have required HTTPS for all pages by adding the [necessary redirect](/docs/domains/#redirect-to-https-and-the-primary-domain), set the HTTP Strict Transport Security (HSTS) header to standardize all client connections on HTTPS and prevent use of HTTP.
-
-Not only does this header help you get an A+ SSL rating from [SSL Labs](https://www.ssllabs.com/ssltest/){.external}, it will help protect your website against protocol downgrade attacks and cookie hijacking.
+If you configure your site to [automatically redirect requests to HTTPS whenever necessary](/docs/domains/#redirect-to-https-and-the-primary-domain), Pantheon will automatically set the HTTP Strict Transport Security (HSTS) header to standardize all client connections on HTTPS and prevent use of HTTP. Either a short (5 minute) or long (366 day) duration may be selected. Using the long duration setting will help you get an A+ SSL rating from [SSL Labs](https://www.ssllabs.com/ssltest/){.external}, and will also help protect your website against protocol downgrade attacks and cookie hijacking.
```
[the remark this answers is not in the document]
Comment (greg-1-anderson) [reply source/_docs/hsts.md]: This is explained at the page linked to above. If `https` is selected, then the short duration is used. If `hsts` is selected, then the longer duration is used. If this is confusing, then perhaps this description could be shortened somewhat.

**Comment by alexfornuto** [comment]
@greg-1-anderson I did a fair bit of rewriting, and moved some information into a partial file to duplicate across docs. Please take a look. :smile:

**Comment by alexfornuto** [comment]
@alexdicianu there was a warning in the HSTS doc, removed by this PR, that said:

>Before adding HSTS to your site, you should review and understand the configuration options that are available. A HSTS header that is configured incompletely or not securely enough diminishes the security protection that HSTS provides.

What do you think about including this warning in the partial file (so both docs will get it), and amending it to say something like:

> Once the HSTS header is set, it cannot be removed. Make _sure_ your HTTPS configuration is correct before enabling.

?

**Comment by greg-1-anderson** [comment]
@alexfornuto The general idea of that suggestion is good; however, some further explanation about what the actual behavior is might be in order.

Here's what actually happens:

- If the customer sets the hsts scheme, their site will start sending 1-year durations on their hsts header
- Any browser that views their site and sees this header will refuse to connect to their site via http for the next year
- If the user changes from the hsts scheme to the https or "any" scheme, then any browser that did NOT visit the site while the hsts scheme was set may still try to connect using http. (If the scheme is https, then that browser will be redirected to https.)

It's a good idea to tell them to make sure their https configuration is correct before enabling, but it is not literally true that they cannot remove the setting. They can remove it, but the removal won't be effective for some portion of their site visitors.

I'm not sure how to best deliver this warning succinctly.

**Comment by alexdicianu** [comment]
How about something like (it may need some rewriting, but the main idea should be this):
```
*Warning*
Once enabled, HSTS headers are cached by browsers for the duration of the max-age period (one year on Pantheon). If at any time you wish to disable https and go back to plain http, users who have already visited your site will receive a browser warning preventing them from accessing the site until the max-age expires or they have manually cleared their browser history related to your site.
```

**Comment by greg-1-anderson** [comment]
@alexfornuto I made a rough attempt at wordsmithing the hsts warning. Please review.

Note also that this feature has now been deployed to production.

**Comment by greg-1-anderson** [comment]
We are considering the current release of this feature to be a DARK DEPLOY. The issue is that we feel that the "include subdomains" feature is dangerous, since it may cause a lasting impact on a site that is not hosted on Pantheon.

We will revisit and update this documentation once we determine how we want to handle the "include subdomains" feature, whether it is to take it out, move it to another selector, or some other solution.

**Comment by alexfornuto** [comment]
@greg-1-anderson pinging to see if there's any update on this.

**Comment by greg-1-anderson** [comment]
Yeah, we are going to soon make `hsts` a separate setting in pantheon.yml. I don't know if `scheme` will be renamed. @ari-gold ?

**Comment by alexfornuto** [comment]
Update for tracking.

File: source/_docs/pantheon-yml.md [hunk 3]
```diff
@@ -37,15 +37,17 @@ protected_web_paths:
 * Wait a few seconds for changes to take effect
 
 ### HTTPS Redirects with HSTS Headers
-Ensure that your site will always use HTTPS to deliver content with `scheme`. Three values are available:
+Ensure that your site will always use HTTPS to deliver content with `enforce_https`. Three values are available:
```
Comment (greg-1-anderson) [inline source/_docs/pantheon-yml.md]: Five values are available.

File: source/_docs/hsts.md (hunk 1 above)
Comment (greg-1-anderson) [reply source/_docs/hsts.md]: 👍 on leaving this in for at least a while.

File: source/_docs/pantheon-yml.md [hunk 4]
```diff
@@ -36,6 +36,27 @@ protected_web_paths:
...
+### HTTPS Redirects with HSTS Headers
+Ensure that your site will always use HTTPS to deliver content with `enforce_https`. Three values are available:
+
+- `off` (default: no https redirect)
+- `transitional` (Redirects to https and includes an HSTS header with a 5-minute duration)
+- `transitional+subdomains` (Redirects to https and includes an HSTS header with a 5-minute duration, including subdomains)
+- `full` (Redirects to https and includes an HSTS header with a 366-day duration)
+- `full+subdomains` (Redirects to https and includes an HSTS header with a 366-day duration, including subdomains)
+
+The preferred `hsts` configuration is shown below:
```
Comment (greg-1-anderson) [inline source/_docs/pantheon-yml.md]: Maybe somewhere around here we should put in a warning that `+subdomains` should not be used if the customer has any subdomains that are not hosted on Pantheon, unless they are sure they will behave correctly with these hsts settings.

File: source/_partials/content/hsts.html [hunk 5]
```diff
@@ -0,0 +1,45 @@
+Using the <a href="/docs/pantheon-yml/"><code>pantheon.yml</code></a> file, you can set the use and duration of the HSTS header, and its affect on subdomains. You can set a short (5 minute) duration header that affects just the domain, or a long (366 day) duration that also applies to subdomains. Using the long duration setting will help you get an A+ SSL rating from <a href="https://www.ssllabs.com/ssltest/" class="external">SSL Labs</a>.
+
+Use of the HSTS header is defined by the <code>scheme</code> directive, and takes three possible values:
```
Comment (greg-1-anderson) [inline source/_partials/content/hsts.html]: s/scheme/enforce_https/

Comment (greg-1-anderson) [reply source/_partials/content/hsts.html]: ... and five possible values.

File: source/_docs/pantheon-yml.md (hunk 4 above)
Comment (helen-yan) [reply source/_docs/pantheon-yml.md]: added below the "preferred hsts config"!

**Comment by alexfornuto** [comment]
I set the PR to merge to the new `gatsby` branch - once y'all are done tinkering with it I'll take a pass at rebasing it to resolve conflicts.

File: source/_docs/http-to-https.md [hunk 6]
```diff
@@ -117,9 +117,11 @@ At this point, all visitors to the site should be able to securely access all pa
 
 ## Redirect to HTTPS and the primary domain
 
-As part of best security practices, we suggest you [Require HTTPS with the HSTS Header](/docs/hsts), as well as a redirect.
+As part of best security practices, we suggest you [Require HTTPS with the HSTS Header](/docs/hsts/), as well as redirect all HTTP traffic to HTTPS.
```
Comment (greg-1-anderson) [inline source/_docs/http-to-https.md]: Wordsmithing: Maybe this should say "which has the effect of redirecting all HTTP traffic to HTTPS", as it is not a separate operation to redirect HTTP traffic to HTTPS but a consequence of setting the HSTS header.

File: source/_docs/pantheon-yml.md [hunk 7]
```diff
@@ -36,6 +36,29 @@ protected_web_paths:
...
+```yaml
+enforce_https: full+subdomains
+```
+
+Note that any option with `+subdomains` should not be used if you have subdomains that are not hosted on Pantheon, unless you are sure they will behave correctly with these hsts settings.
```
Comment (greg-1-anderson) [inline source/_docs/pantheon-yml.md]: We should be more specific about what it means to behave correctly with hsts settings. Their non-Pantheon site needs to be able to respond to https traffic for all requests.

Comment (greg-1-anderson) [reply source/_docs/pantheon-yml.md]: Ah, I see this is explained below, so "as explained below" might be a sufficient addition here.

File: source/_partials/content/hsts.html [hunk 8]
```diff
@@ -0,0 +1,45 @@
+Using the <a href="/docs/pantheon-yml/"><code>pantheon.yml</code></a> file, you can set the use and duration of the HSTS header, and its affect on subdomains. You can set a short (5 minute) duration header, or a long (366 day) duration, and choose whether to affect subdomains. Using the long duration setting will help you get an A+ SSL rating from <a href="https://www.ssllabs.com/ssltest/" class="external">SSL Labs</a>.
+
+Use of the HSTS header is defined by the <code>enforce_https</code> directive, and takes five possible values:
+
+<ul>
+<li>No redirect and no HSTS header (default):
+
+    <pre><code class="hljs less">enforce_https: no</code></pre></li>
```
Comment (greg-1-anderson) [inline source/_partials/content/hsts.html]: This should be "off"

File: source/_docs/pantheon-yml.md [hunk 9]
```diff
@@ -36,6 +36,29 @@ protected_web_paths:
 * You may not be able to protect files or paths with special characters
 * Wait a few seconds for changes to take effect
 
+### HTTPS Redirects with HSTS Headers
+Ensure that your site will always use HTTPS to deliver content with `enforce_https`. Five values are available:
+
+- `no` (default: no https redirect)
```
Comment (greg-1-anderson) [inline source/_docs/pantheon-yml.md]: "off"

**Pull request closed without merging**